upup/pkg/kutil/delete_cluster_vsphere.go

```diff
@@ -17,10 +17,136 @@ limitations under the License.
 package kutil
 
 import (
-	"fmt"
+	"context"
+	"github.com/golang/glog"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/vsphere"
 )
 
+const (
+	typeVM = "VM"
+)
+
+type clusterDiscoveryVSphere struct {
+	cloud        fi.Cloud
+	vsphereCloud *vsphere.VSphereCloud
+	clusterName  string
+}
+
+type vsphereListFn func() ([]*ResourceTracker, error)
+
 func (c *AwsCluster) listResourcesVSphere() (map[string]*ResourceTracker, error) {
-	fmt.Print("listResourcesVSphere to be implemented")
-	return nil, nil
+	vsphereCloud := c.Cloud.(*vsphere.VSphereCloud)
+	if c.Region == "" {
+		c.Region = vsphereCloud.Cluster
+	}
+
+	resources := make(map[string]*ResourceTracker)
+
+	d := &clusterDiscoveryVSphere{
+		cloud:        c.Cloud,
+		vsphereCloud: vsphereCloud,
+		clusterName:  c.ClusterName,
+	}
+
+	listFunctions := []vsphereListFn{
+		d.listVMs,
+	}
+
+	for _, fn := range listFunctions {
+		trackers, err := fn()
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range trackers {
+			resources[t.Type+":"+t.ID] = t
+		}
+	}
+
+	for k, t := range resources {
+		if t.done {
+			delete(resources, k)
+		}
+	}
+	return resources, nil
+}
+
+func (d *clusterDiscoveryVSphere) listVMs() ([]*ResourceTracker, error) {
+	c := d.vsphereCloud
+
+	var trackers []*ResourceTracker
+
+	regexForMasterVMs := "*" + "." + "masters" + "." + d.clusterName + "*"
+	regexForNodeVMs := "nodes" + "." + d.clusterName + "*"
+
+	vms, err := c.VirtualMachines([]string{regexForMasterVMs, regexForNodeVMs})
+	if err != nil {
+		if _, ok := err.(*find.NotFoundError); !ok {
+			return nil, err
+		}
+		glog.Warning(err)
+	}
+
+	for _, vm := range vms {
+		tracker := &ResourceTracker{
+			Name:    vm.Name(),
+			ID:      vm.Name(),
+			Type:    typeVM,
+			deleter: deleteVM,
+			Dumper:  DumpVM,
+			obj:     vm,
+		}
+		trackers = append(trackers, tracker)
+	}
+	return trackers, nil
+}
+
+func deleteVM(cloud fi.Cloud, r *ResourceTracker) error {
+	vsphereCloud := cloud.(*vsphere.VSphereCloud)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	vm := r.obj.(*object.VirtualMachine)
+
+	state, err := vm.PowerState(ctx)
+	if err != nil {
+		return err
+	}
+
+	if state == types.VirtualMachinePowerStatePoweredOn {
+		task, err := vm.PowerOff(context.TODO())
+		if err != nil {
+			return err
+		}
+
+		err = task.Wait(context.TODO())
+		if err != nil {
+			return err
+		}
+	}
+
+	vsphereCloud.DeleteCloudInitISO(fi.String(vm.Name()))
+
+	task, err := vm.Destroy(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	err = task.Wait(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func DumpVM(r *ResourceTracker) (interface{}, error) {
```

Comment:
Resource tracker requires this signature ```func (r *ResourceTracker) (interface{}, error)``` for function DumpVM. We are returning multiple values so we need a map or slice to store it. When the function returns this datatype it gets converted into interface type.
Note: any type in golang can be converted into interface type.